clients/src/main/java/org/apache/kafka/clients/consumer/internals/PrototypeAsyncConsumer.java

```diff
@@ -192,71 +321,38 @@ public PrototypeAsyncConsumer(final ConsumerConfig config,
+                final Fetch<K, V> fetch = pollForFetches(timer);
 
-    /**
-     * Set the fetch position to the committed position (if there is one) or reset it using the
-     * offset reset policy the user has configured (if partitions require reset)
-     *
-     * @return true if the operation completed without timing out
-     * @throws org.apache.kafka.common.errors.AuthenticationException if authentication fails. See the exception for more details
-     * @throws NoOffsetForPartitionException                          If no offset is stored for a given partition and no offset reset policy is
-     *                                                                defined
-     */
-    private boolean updateFetchPositionsIfNeeded(final Timer timer) {
-        // If any partitions have been truncated due to a leader change, we need to validate the offsets
-        ValidatePositionsApplicationEvent validatePositionsEvent = new ValidatePositionsApplicationEvent();
-        eventHandler.add(validatePositionsEvent);
+                if (!fetch.isEmpty()) {
+                    sendFetches();
 
-        // If there are any partitions which do not have a valid position and are not
-        // awaiting reset, then we need to fetch committed offsets. We will only do a
-        // coordinator lookup if there are partitions which have missing positions, so
-        // a consumer with manually assigned partitions can avoid a coordinator dependence
-        // by always ensuring that assigned partitions have an initial position.
-        if (isCommittedOffsetsManagementEnabled() && !refreshCommittedOffsetsIfNeeded(timer))
-            return false;
+                    if (fetch.records().isEmpty()) {
+                        log.trace("Returning empty records from `poll()` "
+                                + "since the consumer's position has advanced for at least one topic partition");
+                    }
 
-        // If there are partitions still needing a position and a reset policy is defined,
-        // request reset using the default policy. If no reset strategy is defined and there
-        // are partitions with a missing position, then we will raise a NoOffsetForPartitionException exception.
-        subscriptions.resetInitializingPositions();
+                    return this.interceptors.onConsume(new ConsumerRecords<>(fetch.records()));
+                }
+                // We will wait for retryBackoffMs
+            } while (timer.notExpired());
 
-        // Finally send an asynchronous request to look up and update the positions of any
-        // partitions which are awaiting reset.
-        ResetPositionsApplicationEvent resetPositionsEvent = new ResetPositionsApplicationEvent();
-        eventHandler.add(resetPositionsEvent);
-        return true;
+            return ConsumerRecords.empty();
+        } finally {
+            this.kafkaConsumerMetrics.recordPollEnd(timer.currentTimeMs());
```

Comment:
The `Timer` is created inside `poll()` is passed into the other methods. The only one that updates it directly is `pollForFetches` which updates it only in the `finally` block at the end:

```java
timer.update(pollTimer.currentTimeMs());
```

I double-checked our implementation against `KafkaConsumer` and they seem to match in that regard. `KafkaConsumer` passes the `Timer` into the methods related to the consumer coordinator, but since we don't have that functionality just yet, we don't.